### PR TITLE
Ensure Pre-Nitro Instance-Types Can Boot EFI-enabled AMI

### LINKED ISSUE #107: [BUG] PostBuild.sh lacks the fix for pre-Nitro support
**Describe the bug**

When using EFI-enabled bootstrap instance-types, the resulting EFI-enabled AMIs will fail to produce EC2s that are bootable on pre-Nitro instance types

**Severity**

Completely Broken (No work-around evident)

**To Reproduce**
1. Launch an EFI-enabled build-EC2
2. Install and execute AMIgen8 scripts
3. Create AMI from installed-to disk
4. Attempt to launch one m4.large and one m5/m6i/m7i.large instance type and wait
6. Look at instance-screensho for (failed) m4.large launch and see the on-screen GRUB error

**Expected behavior**

AMIs produced via this tooling can be used to successfully-launch all non-ARM instance-types 

**Deviance Description**

Pre-Nitro EC2s will boot-fail with a GRUB issue

**Screenshots**

Boot-failure screen
image

**Additional context**

**Fix Suggestions**
Apply `grub2-install` fix from AMIgen9 to this project

### PR DESCRIPTION
**Pull request by ferricoxide**
Closes #107

### CODE REVIEW

**Comment by ferricoxide** [comment]
go codebuild go

File: DualMode-GRUBsetup.sh [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+set -euo pipefail
+
+# Re-Install RPMs as necessary
+dnf -y reinstall grub2-pc
+
+# Move /boot/efi/EFI/redhat/grub.cfg as necessary
+[[ -e /boot/efi/EFI/redhat/grub.cfg ]] && \
```
Comment (lorengordon) [inline DualMode-GRUBsetup.sh:8]: Won't this return 1 if the file _doesn't_ exist, and then errexit due to `set -e`? Is that the desired flow?

File: DualMode-GRUBsetup.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+cat << EOF > "${EFI_HOME}/grub.cfg"
+connectefi scsi
+search --no-floppy --fs-uuid --set=dev ${BOOT_UUID}
+set prefix=(\$dev)${GRUB_DIR}
+export \$prefix
+configfile \$prefix/grub.cfg
+EOF
+
+# Clear out stale grub2-efi.cfg file as necessary
+[[ -e /etc/grub2-efi.cfg ]] && rm -f /etc/grub2-efi.cfg
```
Comment (lorengordon) [inline DualMode-GRUBsetup.sh:47]: Here also...

File: DualMode-GRUBsetup.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# Make our /boot-hosted GRUB2 grub.cfg file
+grub2-mkconfig -o /boot/grub2/grub.cfg
+
+# Nuke grubenv file as necessary
+[[ -e /boot/grub2/grubenv ]] && rm -f /boot/grub2/grubenv
+
+# Create fresh grubenv file
+grub2-editenv /boot/grub2/grubenv create
+
+# Populate fresch grubenv file
```
Comment (lorengordon) [inline DualMode-GRUBsetup.sh:20]:
```suggestion
# Populate fresh grubenv file
```

File: DualMode-GRUBsetup.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+while read -r line
+do
+  key="$( echo "$line" | cut -f1 -d'=' )"
+  value="$( echo "$line" | cut -f2- -d'=' )"
+  grub2-editenv /boot/grub2/grubenv set "${key}"="${value}"
+  done <<< "$( grub2-editenv /boot/efi/EFI/redhat/grubenv list )"
```
Comment (lorengordon) [inline DualMode-GRUBsetup.sh:26]: Might be nice to leave a comment on what this loop is doing, and why it's necessary. Feels pretty obsure.

File: DualMode-GRUBsetup.sh (hunk 1 above)
Comment (ferricoxide) [reply DualMode-GRUBsetup.sh:8]: It didn't fail when running, so, didn't trip over it. Lemme add suspenders…

File: DualMode-GRUBsetup.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+[[ -e /boot/efi/EFI/redhat/grubenv ]] && rm -f /boot/efi/EFI/redhat/grubenv
```
[the remark this answers is not in the document]
Comment (ferricoxide) [reply DualMode-GRUBsetup.sh]: The joys of having ganked the file-contents straight from the vendor-doc (and adding cursory safeties)

File: DualMode-GRUBsetup.sh (hunk 4 above)
Comment (ferricoxide) [reply DualMode-GRUBsetup.sh:31]: The line:
```bash
while read -r line
```
Combined with the line:
```bash
done <<< "$( grub2-editenv /boot/efi/EFI/redhat/grubenv list )"
```
Seemed self-explanatory?

**Review by lorengordon — APPROVED** [review]

File: DualMode-GRUBsetup.sh (hunk 4 above)
Comment (lorengordon) [reply DualMode-GRUBsetup.sh]: Lol, it was more the _why_ that wasn't immediately apparent and required some interpretation or prior experience. The new comment helps. The source is for EFI, the target is for BIOS. Got it.

**Review by lorengordon — APPROVED** [review]
approved pending the manual codebuild test

**Comment by ferricoxide** [comment]
> approved pending the manual codebuild test

image

**Pull request merged**